File: reference/js-reference/lxmf.js
// js/lxmf.js — LXMF message pack/unpack.
//
// LXMF packed format (before encryption):
//   destination_hash(16) + source_hash(16) + signature(64) + msgpack(payload)
//
// For single-packet opportunistic delivery, destination_hash is stripped
// (inferred from the RNS packet header). So on-wire LXMF payload is:
//   source_hash(16) + signature(64) + msgpack(payload)
//
// Payload msgpack array: [timestamp, title_bytes, content_bytes, fields_dict]
// Optional 5th element: stamp (proof-of-work, ignored for now)

'use strict';

import { encode as msgpackEncode, decode as msgpackDecode } from '@msgpack/msgpack';
import { Identity, sha256 } from './identity.js';
import { concatBytes } from './announce.js';
import { TRUNCATED_HASHLENGTH, SIGLENGTH } from './reticulum.js';

const DESTINATION_LENGTH = 16;
const SIGNATURE_LENGTH   = 64;

// ---- Unpack an LXMF message (after decryption) -----------------------

// Unpack from the on-wire format (destination_hash already known from RNS header):
//   source_hash(16) + signature(64) + msgpack(payload)
export async function unpackMessage(data, destHash) {
  if (data.length < TRUNCATED_HASHLENGTH + SIGNATURE_LENGTH + 1) {
    throw new Error('LXMF message too short');
  }

  const sourceHash = data.subarray(0, TRUNCATED_HASHLENGTH);
  const signature  = data.subarray(TRUNCATED_HASHLENGTH, TRUNCATED_HASHLENGTH + SIGNATURE_LENGTH);
  const msgpackData = data.subarray(TRUNCATED_HASHLENGTH + SIGNATURE_LENGTH);

  // Decode msgpack payload
  const payload = msgpackDecode(msgpackData);

  if (!Array.isArray(payload) || payload.length < 4) {
    throw new Error('Invalid LXMF payload structure');
  }

  const timestamp = payload[0];
  const title     = decodeField(payload[1]);
  const content   = decodeField(payload[2]);
  const fields    = payload[3] || {};
  const stamp     = payload.length > 4 ? payload[4] : null;

  // Upstream LXMF computes the signature hash over a re-encoded msgpack
  // containing only the first four elements of the payload (the stamp,
  // if present, is explicitly excluded). We build two candidate hashes:
  // one over a re-encoded stamp-stripped payload, and one over the raw
  // on-wire msgpack bytes as a fallback. verifyMessageSignature will
  // try both because some upstream builds append the stamp as a
  // trailing blob AFTER the msgpack instead of as a 5th array element,
  // and the byte-level re-encode can also drift if our msgpack encoder
  // chooses a different numeric width than the sender's.
  const strippedMsgpack = new Uint8Array(msgpackEncode([timestamp, payload[1], payload[2], fields]));
  const hashedStripped = concatBytes([destHash, sourceHash, strippedMsgpack]);
  const hashStripped   = await sha256(hashedStripped);
  const hashedOriginal = concatBytes([destHash, sourceHash, msgpackData]);
  const hashOriginal   = await sha256(hashedOriginal);

  return {
    sourceHash,
    signature,
    timestamp,
    title,
    content,
    fields,
    stamp,
    destHash,
    msgpackData,
    // Primary hashedPart / messageHash preserve the old field names so
    // verifyMessageSignature's first-pass call still works.
    hashedPart: hashedStripped,
    messageHash: hashStripped,
    msgpackForHash: strippedMsgpack,
    // Fallback view for the "no stamp stripping" variant.
    hashedPartOriginal: hashedOriginal,
    messageHashOriginal: hashOriginal,
    payloadElementCount: payload.length,
  };
}

// Unpack an LXMF message received over an established Link.
//
// Unlike opportunistic delivery (which strips the destination hash because
// it is inferred from the RNS packet header), link-delivered LXMF includes
// the full container:
//   destination_hash(16) + source_hash(16) + signature(64) + msgpack(payload)
// The leading destination hash comes from inside the link ciphertext, not
// from the outer packet header (which carries the link_id instead).
export async function unpackLinkMessage(data) {
  if (data.length < 2 * TRUNCATED_HASHLENGTH + SIGNATURE_LENGTH + 1) {
    throw new Error('LXMF link message too short');
  }
  const destHash = data.subarray(0, TRUNCATED_HASHLENGTH);
  const inner = data.subarray(TRUNCATED_HASHLENGTH);
  return unpackMessage(inner, destHash);
}

// Verify LXMF message signature using the sender's public key.
// Tries the stamp-stripped-and-re-encoded view first (upstream LXMF
// spec behavior), and if that fails falls back to signing over the
// raw on-wire msgpack bytes. Returns an object describing which
// variant matched, or {ok: false} if neither did.
export function verifyMessageSignature(message, senderIdentity) {
  const strippedSigned = concatBytes([message.hashedPart, message.messageHash]);
  if (senderIdentity.verify(message.signature, strippedSigned)) {
    return { ok: true, variant: 'stripped' };
  }
  if (message.hashedPartOriginal && message.messageHashOriginal) {
    const originalSigned = concatBytes([message.hashedPartOriginal, message.messageHashOriginal]);
    if (senderIdentity.verify(message.signature, originalSigned)) {
      return { ok: true, variant: 'original' };
    }
  }
  return { ok: false };
}

// ---- Pack an outbound LXMF message -----------------------------------

export async function packMessage(sourceIdentity, destHash, sourceHash, title, content, fields = {}) {
  const titleBytes   = new TextEncoder().encode(title || '');
  const contentBytes = new TextEncoder().encode(content || '');
  const timestamp    = Date.now() / 1000;  // float seconds

  // Msgpack encode payload
  const msgpackData = new Uint8Array(msgpackEncode([timestamp, titleBytes, contentBytes, fields]));

  // Compute message hash
  const hashedPart = concatBytes([destHash, sourceHash, msgpackData]);
  const messageHash = await sha256(hashedPart);

  // Sign: signed_data = hashed_part + message_hash
  const signedData = concatBytes([hashedPart, messageHash]);
  const signature = sourceIdentity.sign(signedData);

  // On-wire format (destination stripped for opportunistic single-packet):
  //   source_hash(16) + signature(64) + msgpack(payload)
  return concatBytes([sourceHash, signature, msgpackData]);
}

// ---- Helpers ---------------------------------------------------------

function decodeField(val) {
  if (val instanceof Uint8Array || val instanceof ArrayBuffer) {
    return new TextDecoder().decode(val);
  }
  if (typeof val === 'string') return val;
  if (val === null || val === undefined) return '';
  return String(val);
}

File: shared/src/commonMain/kotlin/io/github/thatsfguy/reticulum/lxmf/Lxmf.kt
package io.github.thatsfguy.reticulum.lxmf

import io.github.thatsfguy.reticulum.announce.concatBytes
import io.github.thatsfguy.reticulum.codec.MessagePack
import io.github.thatsfguy.reticulum.crypto.CryptoProvider
import io.github.thatsfguy.reticulum.crypto.Identity
import io.github.thatsfguy.reticulum.protocol.SIGLENGTH
import io.github.thatsfguy.reticulum.protocol.TRUNCATED_HASHLENGTH

/**
 * LXMF message pack/unpack. Port of reference/js-reference/lxmf.js.
 *
 * On-wire formats:
 *   Opportunistic (single-packet, decrypted plaintext):
 *     source_hash(16) + signature(64) + msgpack([ts, title, content, fields])
 *
 *   Link-delivered:
 *     dest_hash(16) + source_hash(16) + signature(64) + msgpack([...])
 *
 * IMPORTANT: source_hash is the sender's DESTINATION hash, not their identity
 * hash. See CLAUDE.md "Key bugs" §3.
 *
 * Signature semantics match upstream LXMF:
 *   hashed_part   = dest_hash + source_hash + msgpack_payload
 *   message_hash  = SHA-256(hashed_part)
 *   signed_data   = hashed_part + message_hash
 *   signature     = Ed25519(sender.sigPriv, signed_data)
 *
 * The sender may emit msgpack bytes that differ from what our encoder
 * produces for the same logical data (different numeric widths, etc.). To
 * handle this we keep BOTH a re-encoded "stripped" view and the raw
 * on-wire bytes; verifyMessageSignature() tries both. CLAUDE.md "Key
 * bugs" §5.
 */
data class LxmfMessage(
    val sourceHash: ByteArray,    // 16 bytes — sender's DESTINATION hash
    val destHash: ByteArray,      // 16 bytes — recipient's destination hash
    val signature: ByteArray,     // 64 bytes
    val timestamp: Double,        // seconds since epoch (may be small if sender clockless)
    val title: String,
    val content: String,
    val fields: Map<Any?, Any?>,
    val stamp: Any?,              // optional 5th payload element
    val msgpackData: ByteArray,   // raw on-wire msgpack bytes
    val msgpackForHash: ByteArray, // re-encoded msgpack (for §5.6 stripped-signature variant)
    val msgpackForId: ByteArray,  // message_id material per §5.5/§5.7.1 (see unpackMessage)
    val payloadElementCount: Int,
)

private const val DESTINATION_LENGTH = TRUNCATED_HASHLENGTH

/**
 * Unpack an opportunistic LXMF message (decrypted plaintext).
 *
 * @param plaintext source_hash(16) + signature(64) + msgpack(...)
 * @param ourDestHash recipient's destination hash (this client's, from RNS header)
 */
suspend fun unpackMessage(
    plaintext: ByteArray,
    ourDestHash: ByteArray,
    crypto: CryptoProvider,
): LxmfMessage {
    require(plaintext.size >= TRUNCATED_HASHLENGTH + SIGLENGTH + 1) { "LXMF message too short" }

    val sourceHash  = plaintext.copyOfRange(0, TRUNCATED_HASHLENGTH)
    val signature   = plaintext.copyOfRange(TRUNCATED_HASHLENGTH, TRUNCATED_HASHLENGTH + SIGLENGTH)
    val msgpackData = plaintext.copyOfRange(TRUNCATED_HASHLENGTH + SIGLENGTH, plaintext.size)

    val payload = MessagePack.decode(msgpackData)
    require(payload is List<*> && payload.size >= 4) { "Invalid LXMF payload structure" }

    val timestamp = (payload[0] as? Number)?.toDouble() ?: 0.0
    val title     = decodeField(payload[1])
    val content   = decodeField(payload[2])
    @Suppress("UNCHECKED_CAST")
    val fields    = (payload[3] as? Map<Any?, Any?>) ?: emptyMap()
    val stamp     = if (payload.size > 4) payload[4] else null

    // Re-encode the first 4 elements for the "stripped" hash variant. The
    // sender excludes the optional stamp from the signed data, so we always
    // hash without it. We pass the ORIGINAL payload[1] / payload[2] (which
    // are ByteArrays per the LXMF wire format) so the re-encoded msgpack
    // is bin-typed, not str-typed.
    val strippedMsgpack = MessagePack.encode(listOf(timestamp, payload[1], payload[2], fields))

    // message_id material per SPEC §5.5 / §5.7.1:
    //   message_id = SHA256(dest_hash || source_hash || msgpack_payload)
    // For an un-stamped message (4 elements) `msgpack_payload` is the
    // payload EXACTLY AS RECEIVED — re-encoding it would diverge from
    // every spec-compliant peer (and fwdsvc's rewrite-cache key) the
    // moment the sender's msgpack encoder drifts from ours (§5.6.1).
    // For a stamped message (5 elements) you cannot slice a 4-element
    // array out of the wire bytes — §5.7.1 mandates dropping element
    // [4] and re-packing the first 4 canonically, which is exactly
    // `strippedMsgpack`. NOT to be confused with `msgpackForHash`,
    // which serves the §5.6 dual-variant signature check.
    val msgpackForId = if (payload.size == 4) msgpackData else strippedMsgpack

    return LxmfMessage(
        sourceHash = sourceHash,
        destHash = ourDestHash,
        signature = signature,
        timestamp = timestamp,
        title = title,
        content = content,
        fields = fields,
        stamp = stamp,
        msgpackData = msgpackData,
        msgpackForHash = strippedMsgpack,
        msgpackForId = msgpackForId,
        payloadElementCount = payload.size,
    )
}

/**
 * Unpack a link-delivered LXMF message (received decrypted from a Link).
 *
 * Format: dest_hash(16) + source_hash(16) + signature(64) + msgpack(...)
 */
suspend fun unpackLinkMessage(data: ByteArray, crypto: CryptoProvider): LxmfMessage {
    require(data.size >= 2 * TRUNCATED_HASHLENGTH + SIGLENGTH + 1) { "LXMF link message too short" }
    val destHash = data.copyOfRange(0, TRUNCATED_HASHLENGTH)
    val inner    = data.copyOfRange(TRUNCATED_HASHLENGTH, data.size)
    return unpackMessage(inner, destHash, crypto)
}

/**
 * Pack an outbound opportunistic LXMF message.
 *
 * Returns the plaintext bytes (source_hash + signature + msgpack) ready to
 * be Token-encrypted to the recipient's encryption pub key.
 */
suspend fun packMessage(
    sourceIdentity: Identity,
    destHash: ByteArray,
    sourceHash: ByteArray,
    title: String,
    content: String,
    timestampSeconds: Double,
    fields: Map<Any?, Any?> = emptyMap(),
    crypto: CryptoProvider,
    /** Optional LXMF stamp per SPEC §5.7. When non-null, this is the
     *  32-byte PoW value the caller computed via [LxmfStamp.findStamp];
     *  it's appended as element [4] of the wire msgpack array. The
     *  Ed25519 signature is still computed over the 4-element
     *  "stripped" variant (§5.6) so the stamp can be added/removed
     *  without invalidating the sig. */
    stamp: ByteArray? = null,
): ByteArray {
    val titleBytes   = title.encodeToByteArray()
    val contentBytes = content.encodeToByteArray()
    // Sign over the 4-element form per §5.6. The wire form below may
    // additionally carry a 5th element (stamp) — receivers strip it
    // before verifying.
    val signableMsgpack = MessagePack.encode(listOf(timestampSeconds, titleBytes, contentBytes, fields))

    val hashedPart  = concatBytes(listOf(destHash, sourceHash, signableMsgpack))
    val messageHash = crypto.sha256(hashedPart)
    val signedData  = concatBytes(listOf(hashedPart, messageHash))
    val signature   = sourceIdentity.sign(signedData)

    val wireMsgpack = if (stamp != null) {
        MessagePack.encode(listOf(timestampSeconds, titleBytes, contentBytes, fields, stamp))
    } else {
        signableMsgpack
    }
    return concatBytes(listOf(sourceHash, signature, wireMsgpack))
}

/**
 * Pack an outbound link-delivered LXMF message.
 *
 * Wire format: dest_hash(16) || source_hash(16) || signature(64) || msgpack(...).
 * Signature semantics are identical to the opportunistic form (the
 * sender signs SHA-256(dest_hash || source_hash || msgpack || hashed_part_hash)),
 * so we just compose: prepend dest_hash to packMessage's output.
 *
 * Returns the plaintext bytes ready to be Token-encrypted with the
 * link's session key (link.derivedKey).
 */
suspend fun packLinkMessage(
    sourceIdentity: Identity,
    destHash: ByteArray,
    sourceHash: ByteArray,
    title: String,
    content: String,
    timestampSeconds: Double,
    fields: Map<Any?, Any?> = emptyMap(),
    crypto: CryptoProvider,
    /** Optional LXMF stamp per SPEC §5.7 — see [packMessage]. */
    stamp: ByteArray? = null,
): ByteArray {
    val opportunisticBody = packMessage(
        sourceIdentity   = sourceIdentity,
        destHash         = destHash,
        sourceHash       = sourceHash,
        title            = title,
        content          = content,
        timestampSeconds = timestampSeconds,
        fields           = fields,
        crypto           = crypto,
        stamp            = stamp,
    )
    return concatBytes(listOf(destHash, opportunisticBody))
}

/**
 * Verify an LXMF message signature using the sender's public key.
 *
 * Tries the stripped (re-encoded msgpack) variant first, then falls back to
 * the raw on-wire bytes. Returns the matched variant or null if neither
 * passes.
 */
suspend fun verifyMessageSignature(
    msg: LxmfMessage,
    senderIdentity: Identity,
    crypto: CryptoProvider,
): SignatureVariant? {
    val strippedHashedPart = concatBytes(listOf(msg.destHash, msg.sourceHash, msg.msgpackForHash))
    val strippedHash = crypto.sha256(strippedHashedPart)
    val strippedSigned = concatBytes(listOf(strippedHashedPart, strippedHash))
    if (senderIdentity.verify(msg.signature, strippedSigned)) return SignatureVariant.STRIPPED

    val originalHashedPart = concatBytes(listOf(msg.destHash, msg.sourceHash, msg.msgpackData))
    val originalHash = crypto.sha256(originalHashedPart)
    val originalSigned = concatBytes(listOf(originalHashedPart, originalHash))
    if (senderIdentity.verify(msg.signature, originalSigned)) return SignatureVariant.ORIGINAL

    return null
}

enum class SignatureVariant { STRIPPED, ORIGINAL }

private fun decodeField(v: Any?): String = when (v) {
    null         -> ""
    is ByteArray -> v.decodeToString()
    is String    -> v
    else         -> v.toString()
}
